### ISSUE TITLE
The way logind in preallocates VT and spawns agetty on the active tty is racy

### ISSUE THREAD

**Issue by fbuihuu** [systemd/systemd#23660]
While logind is started it starts "seat0" by:

1. pre-allocating VTs (`seat_preallocate_vts()`)
2. reading the active VT (`seat_read_active_vt()`)
3. spawning `getty@.service` on the active VT if it's not busy (`manager_spawn_autovt()`)

I don't know what `seat_preallocate_vts()` is supposed to do exactly however it consists essentially in opening and closing each tty device. Later (in step #3) `manager_spawn_autovt()` checks whether the tty device is currently busy by retrieving the tty busy mask with `ioctl(fd, VT_GETSTATE, ..)` on "/dev/tty1.

The problem here is that closing a tty is done (via a workqueue) asynchronously by the kernel. Hence even if no process is using `tty2` (for instance), the fact that logind just opened and closed this device (in `seat_preallocate_vts()`) can make the device still busy at the time its state is checked by `manager_spawn_autovt()`. Hence this "race" can result in a missing login prompt on the active tty.

This problem might be reproduced by switching to the rescue mode (`systemctl rescue`) so logind is stopped and then changing the current VT and switching back to the default target (`chvt 2; systemctl default`).

Note that this problem probably doesn't occur on regular boots because `getty@tty1.service` is usually enabled hence this unit is started by PID1 regardless of whether logind finds tty1 busy or not.

**Comment by poettering** [comment]
> The problem here is that closing a tty is done (via a workqueue) asynchronously by the kernel. Hence even if no process is using `tty2` (for instance), the fact that logind just opened and closed this device (in `seat_preallocate_vts()`) can make the device still busy at the time its state is checked by `manager_spawn_autovt()`. Hence this "race" can result in a missing login prompt on the active tty.

smells like a kernel bug no? i.e. should synchronize on the workqueues before reporting the VT_GETSTATE data.

**Comment by fbuihuu** [comment]
I would tend to agree however it was not  clear to me what `seat_preallocate_vts()` is actually used for as it consists essentially in opening and closing each tty device. Any idea ?

**Comment by fbuihuu** [comment]
Anyway, I'll check with our kernel guys but for now let's close this report. I'll update it  as soon as I get more info.

**Comment by poettering** [comment]
> I would tend to agree however it was not clear to me what `seat_preallocate_vts()` is actually used for as it consists essentially in opening and closing each tty device. Any idea ?

This allocates the VTs, in the sense that you can actually switch to them. You cannot switch to VTs that have never been opened. Hence we open them, so that A-F2, A-F3, A-F4, … and so on actually work.

**Comment by fbuihuu** [comment]
@poettering let me re-open this bug because while the interface provided by the kernel seems boggus indeed there's still something debatable in the way logind try to make use of it.

Indeed logind checks whether the current tty is busy or not (done by `vt_is_busy()` in `manager_spawn_autovt()`) before spawning agetty on the tty, however the whole sequence is and will always be racy anyways since nothing guaranty that another process will not take the tty between `vt_is_busy()` and `agetty`.

If that's correcy, why not simply skipping the test and always spawn agetty letting it deal with any kind of error and letting it decide whether it should wait or not ?

**Comment by poettering** [comment]
> The thing is that I made `seat_preallocate_vts()` a NOP and this didn't have any side effect, I was still able to switch to the different ttys...

probably because of the VT logic having those VTs set up for you already? but we don't rely on that.

**Comment by poettering** [comment]
> Indeed logind checks whether the current tty is busy or not (done by `vt_is_busy()` in `manager_spawn_autovt()`) before spawning agetty on the tty, however the whole sequence is and will always be racy anyways since nothing guaranty that another process will not take the tty between `vt_is_busy()` and `agetty`.

Well, it's the only API there is... The whole VT subsystem is a racy mess. We use the mess there is, not the race-free subsystem that doesn't exist...

> If that's correcy, why not simply skipping the test and always spawn agetty letting it deal with any kind of error and letting it decide whether it should wait or not ?

that's going to look terrible in the logs.

**Comment by poettering** [comment]
also, it is not as racy IRL, because code usually opens the VT first, and then switches to it. Thus, when logind looks at it because the VT was just switched to the VT is generally already opened.

**Comment by fbuihuu** [comment]
> probably because of the VT logic having those VTs set up for you already? but we don't rely on that.

Sorry but I don't know what you meant by "VT logic" here.

**Comment by fbuihuu** [comment]
> Well, it's the only API there is... The whole VT subsystem is a racy mess. We use the mess there is, not the race-free subsystem that doesn't exist...

I don't get the point here. If the API is broken at least for what logind is trying to ensure, there's no point in using the API in the first place.

**Comment by fbuihuu** [comment]
> that's going to look terrible in the logs.

Hmm, why ? do you expect that the target tty busy ?

**Comment by poettering** [comment]
> Sorry but I don't know what you meant by "VT logic" here.

I wanted to say "VT setup logic", i.e. `systemd-vconsole-setup.service`

**Comment by poettering** [comment]
> I don't get the point here. If the API is broken at least for what logind is trying to ensure, there's no point in using the API in the first place.

is it broken though? it's racy, but it should generally work.

**Comment by poettering** [comment]
> Hmm, why ? do you expect that the target tty busy ?

yes, think X11. whenever you switch to a VT with X11 on we'd end up starting a getty on it, which will fail, because the VT is busy.
